requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+- Retry logic per Activity and Workflow
+- Timeout logic per activity and workflow
+- Exactly-once semantics (!!)
+- Go, [Python](https://docs.temporal.io/kb/python-sandbox-environment), and Typescript libraries that support deterministic code execution and retries (!!)
+- Composability of Activities and Workflows
+- Dashboard for observing Workflow status
+
+Finally, Temporal is a good fit because we are a small team. We should be focusing on the business of building our product. Nothing about what we are building that would be run on Temporal is novel enough to justify us to invest in a competing runtime.
+
+### Who uses it?
+
+- **Uber** (original developer - originally called [Cadence](https://github.com/uber/cadence))
+- **Coinbase**
+- **Doordash**
+- **Instacart**
+- **Hashicorp**
+- **Netflix**
+- **Snapchat**
+- **Stripe**
+- **Datadog**
+- **Box**
+- **Airbyte**
+- A whole lot more…
+
+### Quick Concept Overview
+
+A workflow in [Temporal.io](http://temporal.io/) is a set of simple, easy-to-understand steps that describe a complex process. Workflows are made up of activities, which are individual steps in the workflow that perform specific tasks. Activities can be executed in parallel, and Temporal provides retry logic, timeout logic, and exactly-once semantics. Workflows and activities can be composed together, and Temporal provides a dashboard for observing workflow status.
+
+# Example Workflow
+
+We currently use it for our manual backfills to S3 for certain clients. You can see an example workflow here:
+
+[https://github.com/posthog/energize](https://github.com/posthog/energize)
+
+Note: that this will be folded into the Posthog/Posthog repo. For now this is separate for example purposes.
+
+The specific workflow for S3 exporting can be found [here](https://github.com/PostHog/energize/blob/main/src/energize/workflows/exports/s3.py).
+
+# Temporal Cloud
+
+Why ☁️?
+
+There is one critical piece of the Temporal topology that causes loss if it goes down: The scheduler. As long as the scheduler is up. When you schedule a job - it will be completed. If the scheduler is down for an extended period of time there is a chance that the requesting service would drop the intent and that workflow would never be run. As long as the scheduler is up and we are with the tunable TTL of the workflow, as soon as a worker comes back online we will process that workflow.
+
+Luckily [Temporal.io](http://Temporal.io) has a hosted version of their scheduler where we can just submit our workflows and connect our workers to and the rest should be relatively smooth sailing. We only have to worry about writing the workflows themselves and keeping the workers happy.
+
+# Ownership
+
+This is why I’ve wanted to get you all here. Here is how ownership here would most likely shake out:
+
+**Infrastructure**
+
+- K8s Deployment of the workers that execute Workflows and Actions
+- CD of these workers
```

Comment:
I would be interested to understand more how does Temporal's worker abstraction map to K8s: is it image with a long-running process that we just spin up? Is it more on-demand (like how Airflow uses `KubernetesOperator`)?

Reply:
Also, I feel this responsibility will likely be shared with other teams. At least when it comes to the determination and updating of worker requirements.

Reply:
> is it image with a long-running process that we just spin up

Afaik yes, you basically compile your workflows + activities into an executable, which then uses the temporal SDK to pull tasks + run your code

Reply:
Does that mean there is no "hot reload" of workflows? We do need to re-compile with every workflow change?

Reply:
@tomasfarias my understanding is that any pod can be a worker: they just open a grpc connection to the scheduler, advertise their capabilities and wait for incoming tasks.

While the heaviest tasks will be new dedicated services, we could also add temporal capabilities to the existing services, when that makes sense.